rust/signed_doc/src/lib.rs

```diff
@@ -1 +1,192 @@
+}
+
+impl Display for CatalystSignedDocument {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        writeln!(f, "{}", self.inner.metadata)?;
+        writeln!(f, "JSON Payload {:#}\n", self.inner.payload)?;
+        writeln!(f, "Signatures [")?;
+        for signature in &self.inner.signatures {
+            writeln!(f, "  0x{:#}", hex::encode(signature.signature.as_slice()))?;
+        }
+        writeln!(f, "]\n")?;
+        writeln!(f, "Content Errors [")?;
+        for error in &self.inner.content_errors {
+            writeln!(f, "  {error:#}")?;
+        }
+        writeln!(f, "]")
+    }
+}
+
+#[derive(Default)]
+/// Inner type that holds the Catalyst Signed Document with parsing errors.
+struct InnerCatalystSignedDocument {
+    /// Document Metadata
+    metadata: Metadata,
+    /// JSON Payload
+    payload: serde_json::Value,
+    /// Signatures
+    signatures: Vec<coset::CoseSignature>,
+    /// Raw COSE Sign bytes
+    cose_sign: coset::CoseSign,
+    /// Content Errors found when parsing the Document
+    content_errors: Vec<String>,
+}
+
+// Do this instead of `new`  if we are converting a single parameter into a struct/type we
+// should use either `From` or `TryFrom` and reserve `new` for cases where we need
+// multiple parameters to actually create the type.  This is much more elegant to use this
+// way, in code.
+impl TryFrom<Vec<u8>> for CatalystSignedDocument {
+    type Error = anyhow::Error;
+
+    fn try_from(cose_bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        // Try reading as a tagged COSE SIGN, otherwise try reading as untagged.
+        let cose = coset::CoseSign::from_tagged_slice(&cose_bytes)
+            .or(coset::CoseSign::from_slice(&cose_bytes))
+            .map_err(|e| anyhow::anyhow!("Invalid COSE Sign document: {e}"))?;
+        let mut content_errors = Vec::new();
+        let expected_header = cose_protected_header();
+
+        if cose.protected.header.content_type != expected_header.content_type {
+            content_errors
+                .push("Invalid COSE document protected header `content-type` field".to_string());
+        }
+
+        if !cose.protected.header.rest.iter().any(|(key, value)| {
+            key == &coset::Label::Text(CONTENT_ENCODING_KEY.to_string())
+                && value == &coset::cbor::Value::Text(CONTENT_ENCODING_VALUE.to_string())
+        }) {
+            content_errors.push(
+                "Invalid COSE document protected header {CONTENT_ENCODING_KEY} field".to_string(),
+            );
+        }
+        let metadata = Metadata::from(&cose.protected);
+        if metadata.has_error() {
+            content_errors.extend_from_slice(metadata.content_errors());
+        }
+        let payload = match &cose.payload {
+            Some(payload) => {
+                let mut buf = Vec::new();
+                let mut bytes = payload.as_slice();
+                brotli::BrotliDecompress(&mut bytes, &mut buf)?;
```

Comment:
Here we should handle the `encoding` from the header.  IF its an encoding we support, we decode accordingly (such as decompress here).  We should be able to support `br` and None (Raw).  Any other content encoding is currently undefined, and this is where the error would be generated.  Not when its extracted from the metadata.